src/combinatorics/set_family/lym.lean

```diff
@@ -0,0 +1,251 @@
+
+/-- First set of the double counting. Effectively `{(s, t) | s ∈ 𝒜, t ∈ s.image (erase s)}`. -/
+def lym_above (𝒜 : finset (finset α)) : finset (finset α × finset α) :=
+𝒜.sup $ λ s, s.image $ λ x, (s, erase s x)
+
+/-- For each `s ∈ 𝒜` there are `r` possible `t` to make an element of `lym_above`. -/
+lemma _root_.set.sized.card_lym_above (h𝒜 : (𝒜 : set (finset α)).sized r) :
+  (lym_above 𝒜).card = 𝒜.card * r :=
+begin
+  rw [lym_above, sup_eq_bUnion, card_bUnion],
+  { convert sum_const_nat _,
+    refine λ x hx, (card_image_of_inj_on $ λ a ha b hb h, _).trans (h𝒜 hx),
+    exact x.erase_inj_on ha hb (prod.mk.inj h).2 },
+  { simp only [disjoint_left, mem_image],
+    rintro _ _ _ _ h a ⟨_, _, rfl⟩ ⟨_, _, a₂⟩,
+    exact h (prod.mk.inj a₂.symm).1 }
+end
+
+variables [fintype α]
+
+/-- Second set of the double counting. We're trying to get the same set, but we count `t` first, so
+we overestimate a bit. It's pretty much `{(s, t) | t ∈ ∂𝒜, ∃ a ∉ t, s = t ∪ {a}}` -/
+def lym_below (𝒜 : finset (finset α)) : finset (finset α × finset α) :=
+(∂𝒜).sup $ λ t, tᶜ.image $ λ a, (insert a t, t)
+
+/-- For each `t ∈ ∂𝒜`, there are `card α - r + 1` choices for what to add to it to make an element
+of `lym_below`. -/
+lemma _root_.set.sized.card_lym_below (h𝒜 : (𝒜 : set (finset α)).sized r) :
+  (lym_below 𝒜).card = (∂𝒜).card * (fintype.card α - (r - 1)) :=
+begin
+  rw [lym_below, sup_eq_bUnion, card_bUnion],
+  { refine sum_const_nat (λ s hs, _),
+    rw [card_image_of_inj_on, card_compl, h𝒜.shadow hs],
+    intros a ha b hb h,
+    injection h with hab,
+    have q := mem_insert_self a s,
+    rw [hab, mem_insert] at q,
+    exact q.resolve_right (mem_sdiff.1 ha).2 },
+  intros s hs t ht hst,
+  rw disjoint_left,
+  simp_rw [mem_image, not_exists, exists_prop, mem_compl, exists_imp_distrib, prod.forall,
+    prod.mk.inj_iff, and_imp, not_and],
+  rintro _ b i hi rfl rfl j hj k,
+  exact hst.symm,
+end
+
+lemma lym_above_subset_lym_below : lym_above 𝒜 ⊆ lym_below 𝒜 :=
+begin
+  rintro ⟨s, t⟩,
+  simp only [lym_above, lym_below, mem_sup, mem_shadow_iff, true_and, and_imp,
+    exists_prop, mem_sdiff, mem_image, prod.mk.inj_iff, mem_univ, exists_imp_distrib],
+  rintro s hs a hx rfl rfl,
+  exact ⟨s.erase a, ⟨s, hs, a, hx, rfl⟩, a, mem_compl.2 $ not_mem_erase _ _, insert_erase hx, rfl⟩,
+end
+
+/-- The downward **local LYM inequality**. `𝒜` takes up less of `α^(r)` (the finsets of card `r`)
+than `∂𝒜` takes up of `α^(r - 1)`. -/
+lemma local_lym (hr : r ≠ 0) (h𝒜 : (𝒜 : set (finset α)).sized r) :
+  (𝒜.card : 𝕜) / (fintype.card α).choose r ≤ (∂𝒜).card / (fintype.card α).choose (r - 1) :=
+begin
+  obtain hr' | hr' := lt_or_le (fintype.card α) r,
+  { rw [choose_eq_zero_of_lt hr', cast_zero, div_zero],
+    exact div_nonneg (cast_nonneg _) (cast_nonneg _) },
+  { apply lym_aux hr hr',
+    rw [←h𝒜.card_lym_above, ←tsub_tsub_assoc hr' (pos_iff_ne_zero.2 hr), ←h𝒜.card_lym_below],
+    exact card_le_of_subset lym_above_subset_lym_below }
+end
```

Comment:
Maybe this has been discussed before, but why don't we generalize this double counting argument? Given a bipartite graph on vertex sets `L`, `R` (i.e. a relation `L -> R -> Prop` if you like), if every degree in `L` is at least `l` and every degree in `R` is at most `r`, then `l * card L <= r * card R` (proof: one is a lower bound on the number of edges and the other is an upper bound). In particular, for a subset `A : set L`, we have a lower bound `l * card A <= r * card (N A)` for the size of the neighbourhood `N A : set R`. In the case of local LYM, the graph is regular, so `l * card L = r * card R` and hence (by division) the density of `N A` is at least the density of `A`.

This general statement is of interest e.g. because it's often how you prove Hall's condition for a specific bipartite graph.

Reply:
Done. What do you think? (apart from the fact that `lym` is now a mess with all the new lemmas)

Reply:
Looks great!